spec/controllers/users/emails_controller_spec.rb

```diff
@@ -12,17 +12,40 @@
 
   context 'user visits add an email address page' do
     let(:user) { create(:user) }
+
     before do
       stub_sign_in(user)
       stub_analytics
     end
-    it 'renders the index view' do
+
+    it 'renders the show view' do
       get :show
 
       expect(@analytics).to have_logged_event('Add Email Address Page Visited')
     end
   end
 
+  context 'user visits add an email address from SP consent flow' do
+    let(:user) { create(:user) }
+    let(:current_sp) { create(:service_provider) }
+
+    before do
+      stub_sign_in(user)
+      subject.session[:sp] = {
+        issuer: current_sp.issuer,
+        acr_values: Saml::Idp::Constants::IAL1_AUTHN_CONTEXT_CLASSREF,
+        requested_attributes: [:email],
+        request_url: 'http://localhost:3000',
+      }
+    end
+
+    it 'renders the show view with a link back to continue SP consent' do
```

Comment:
Minor: I think typically we may want to separate controller and view spec assertions. We're not expecting the controller to be the one that "renders a link", that's what happens in the view. The behavior we're expecting from the controller is for it to assign an instance variable that can be accessed by the view. I don't think we're really testing that either, since the view isn't using `controller.pending_completions_consent?`, it's using `@pending_completions_consent`.

I think ideally we'd have controller specs asserting the different expected values of `@pending_completions_consent`, and view specs asserted the expected rendered output.

Reply:
That makes a lot of sense. Thanks!